resolwe/flow/migrations/0037_migrate_m2o.py

```diff
@@ -0,0 +1,227 @@
+    duplicate.slug = None
+    duplicate.entity2 = None
+    duplicate.collection = None
+    duplicate.name = 'Copy of {}'.format(original.name)
+    duplicate.duplicated = now()
+
+    duplicate.save(force_insert=True)  # pylint: disable=protected-access
+
+    copy_permissions(original, duplicate, model='data', apps=apps)
+
+    # Override fields that are automatically set on create.
+    duplicate.created = original.created
+    duplicate.save()  # pylint: disable=protected-access
+
+    if original.location:
+        original.location.data.add(duplicate)  # pylint: disable=no-member
+
+    duplicate.storages.set(original.storages.all())  # pylint: disable=no-member
+
+    for migration in original.migration_history.order_by('created'):  # pylint: disable=no-member
+        migration.pk = None
+        migration.data = duplicate
+        migration.save(force_insert=True)
+
+    # Inherit existing child dependencies.
+    DataDependency.objects.bulk_create([
+        DataDependency(child=duplicate, parent=dependency.parent, kind=dependency.kind)
+        for dependency in DataDependency.objects.filter(child=original)
+    ])
+    # Inherit existing parent dependencies.
+    DataDependency.objects.bulk_create([
+        DataDependency(child=dependency.child, parent=duplicate, kind=dependency.kind)
+        for dependency in DataDependency.objects.filter(parent=original)
+    ])
+
+    return duplicate
+
+
+def duplicate_m2m(apps, schema_editor):
+    """Duplicate Data/Entity belonging to more than one Entity/Collection."""
+    Data = apps.get_model('flow', 'Data')
+    Entity = apps.get_model('flow', 'Entity')
+
+    # 1: Duplicate entities that are in multiple collections. This is
+    # done first as entities also contain data that is therefore
+    # handled "for free".
+    for entity in Entity.objects.annotate(collection_count=models.Count('collections')).filter(collection_count__gt=1):
+        collections = entity.collections.all()
+        for i, collection in enumerate(collections):
+            if i == 0:
+                # Entity for one of the collection does not need to be
+                # duplicated (e.g first one) but should be modified to
+                # only belong to that collection.
+                entity.collection = collection
+                entity.save()
+                for datum in entity.data.all():
+                    datum.collection = collection
+                    datum.save()
+                continue
+
+            duplicate = duplicate_entity(entity, apps)
+            duplicate.collection = collection
+            duplicate.save()
+            for datum in duplicate.data.all():
+                datum.collection = collection
+                datum.save()
+
+            entity.collections.remove(collection)
+            for datum in entity.data.all():
+                datum.collection_set.remove(collection)
+
+    # 2: Duplicate data from multiple collections.
+    for data in Data.objects.iterator():
```

Comment:
Is there a reason why you handle entities differently?

Reply:
What do you mean by "handle differently"?